### PR TITLE
feat: add TES models with unit tests

### PR DESCRIPTION
**Pull request by Karanjot786**
Introduce Pydantic models for Task Execution Services (TES) to enhance type safety and data validation, conforming to the GA4GH schemas. Additionally, provide comprehensive unit tests to ensure model integrity and functionality.

**Changes:**

- **Added Models:**
  - `crategen/models/tes_models.py`: Contains all TES-related Pydantic models, including `TESData`, `TESInput`, `TESOutput`, `TESExecutor`, `TESTaskLog`, `TESResources`, `TESExecutorLog`, `TESOutputFileLog`, `TESFileType`, and `TESState`.

- **Added Unit Tests:**
  - `tests/unit/test_tes_models.py`: Includes unit tests for all TES models, covering valid instances, validation errors, and edge cases.

- **Updated Package Initialization:**
  - `crategen/models/__init__.py`: Added imports for TES models and defined `__all__` to specify the public API of the `models` package.
 
**Attribution:**
- Portions of the TES models were adapted from @SalihuDickson's implementations to maintain consistency and ensure alignment with project standards.

**Documentation:**

- Added comprehensive docstrings to all public classes and methods in `tes_models.py` to improve code readability and maintainability.

**Linting and Testing:**

- Resolved all linting errors using Ruff by adding necessary docstrings and replacing magic values with named constants.
- Ensured all unit tests pass successfully.

**Notes:**

- This PR focuses solely on adding TES models and their corresponding unit tests.
- Integration of these models into converters (`tes_converter.py` and `wes_converter.py`) and addition of related tests will be handled in subsequent PRs to maintain manageable and focused code reviews.

## Summary by Sourcery

Add Pydantic models for Task Execution Services (TES) to enhance type safety and data validation, along with comprehensive unit tests to ensure model integrity and functionality. Update package initialization and resolve linting issues.

New Features:
- Introduce Pydantic models for Task Execution Services (TES) to enhance type safety and data validation, conforming to the GA4GH schemas.

Enhancements:
- Update package initialization to include imports for TES models and define `__all__` to specify the public API of the `models` package.

Documentation:
- Add comprehensive docstrings to all public classes and methods in `tes_models.py` to improve code readability and maintainability.

Tests:
- Add comprehensive unit tests for all TES models, covering valid instances, validation errors, and edge cases.

Chores:
- Resolve all linting errors using Ruff by adding necessary docstrings and replacing magic values with named constants.

### CODE REVIEW

**Comment by SalihuDickson** [comment]
@Karanjot786, the generally accepted practice to split PRs like this is to cherry pick the commits. Copying and pasting someones else's code (mine in this case) with no attribution is not very professional.

**Comment by Karanjot786** [comment]
Hi @SalihuDickson,

Thank you for bringing this to my attention and for your understanding.

I acknowledge that I copied your code without proper attribution in the recent PR. My intention was solely to replicate the models accurately since we are working on the same project and ensuring consistency is crucial. I wanted to avoid potential discrepancies that might arise from rewriting the models from scratch.

I understand that copying code without attribution is not professional, regardless of the circumstances. In the future, I will adhere to best practices by cherry-picking commits and making the necessary adjustments to maintain both integrity and clarity in our work. I appreciate your guidance on this matter and will ensure that all contributions are properly attributed moving forward.

Once again, I apologize for any inconvenience this may have caused and thank you for your patience and support.

**Comment by uniqueg** [comment]
> Hi @SalihuDickson,
> 
> Thank you for bringing this to my attention and for your understanding.
> 
> I acknowledge that I copied your code without proper attribution in the recent PR. My intention was solely to replicate the models accurately since we are working on the same project and ensuring consistency is crucial. I wanted to avoid potential discrepancies that might arise from rewriting the models from scratch.
> 
> I understand that copying code without attribution is not professional, regardless of the circumstances. In the future, I will adhere to best practices by cherry-picking commits and making the necessary adjustments to maintain both integrity and clarity in our work. I appreciate your guidance on this matter and will ensure that all contributions are properly attributed moving forward.
> 
> Once again, I apologize for any inconvenience this may have caused and thank you for your patience and support.

Note that since we squash and merge commits in a PR anyways, you can just attribute co-authorship of a commit with [`co-authored-by` directive](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) in the commit message for the squash. Let's just remember to do that when we are good to go.

File: crategen/models/tes_models.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    **Attributes:**
+    - **UNKNOWN** (`str`): The task state is unknown.
+    - **QUEUED** (`str`): The task is queued.
+    - **INITIALIZING** (`str`): The task is initializing.
+    - **RUNNING** (`str`): The task is running.
+    - **PAUSED** (`str`): The task is paused.
+    - **COMPLETE** (`str`): The task is complete.
+    - **EXECUTOR_ERROR** (`str`): The task encountered an executor error.
+    - **SYSTEM_ERROR** (`str`): The task encountered a system error.
+    - **CANCELLED** (`str`): The task was cancelled.
```
Comment (uniqueg) [inline crategen/models/tes_models.py:37]: Please use Google-style docstrings, just like for the other code:
- No emphasis of section and attribute/arg names
- Indicate types in code (type hints), _not_ in docstrings

**Review by uniqueg — DISMISSED** [review]
This looks really good! :)

However, one major thing from looking at the nice summary that Sourcery provided: Are you sure you created the models for [TES v1.1](https://github.com/ga4gh/task-execution-schemas/blob/v1.1/openapi/task_execution_service.openapi.yaml) as previously discussed? There are a few minor inconsistencies with the models, but maybe it's because you have used an older version?

Here's a list of the issues I found:
- `TESState` (from `tesState`):
  - `CANCELING` is missing (added in v1.1)
  - `PREEMPTED` is missing (added in v1.1)
- `TESInput` (from `tesInput`):
  - `type` is optional for creating tasks (requirement dropped in v1.1, I believe), but should be filled in once that information becomes available; probably not enforcable on the validation side
- `TESOutput` (from `tesOutput`):
  - `path_prefix` is missing (added in v1.1); note that it is required if `TESOutput.path` contains wildcards; requires custom Pydantic validator
  - `type` as above
- `TESExecutor` (from `tesExecutor`):
  - `ignore_error` is missing (added in v1.1)
- `TESData` (from `tesTask`):
  - Only `executors` is required, `inputs` and `outputs` are not

Apart from that, a few minor comments:
- Check the docstrings and make sure they are consistent with Google's docstring style and your other docstrings in the codebase (see individual comment for more info)
- Address the open Sourcery comments

File: crategen/models/tes_models.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+    **Attributes:**
+
+    - **logs** (`list[TESExecutorLog]`): Logs for each executor.
+    - **metadata** (`Optional[dict[str, str]]`): Arbitrary logging metadata included by the implementation.
+    - **start_time** (`Optional[datetime]`): When the task started, in RFC 3339 format.
+    - **end_time** (`Optional[datetime]`): When the task ended, in RFC 3339 format.
+    - **outputs** (`list[TESOutputFileLog]`): Information about all output files. Directory outputs are flattened into separate items.
+    - **system_logs** (`Optional[list[str]]`): System logs are any logs the system decides are relevant, which are not tied directly to an Executor process. Content is implementation specific: format, size, etc.
+    - **status** (`Optional[str]`): The status of the task.
```
[the remark this answers is not in the document]
Comment (uniqueg) [reply crategen/models/tes_models.py:233]: Indeed, `state` (not `status`) is a property of `tesTask` (`TESData` in your language), not `TESTaskLog` - so please remove from the docstring.

BTW: Also note that, again, `state` is, of course, implicitly required when requesting a task resource from the service, but it is not allowed to be set by the user. Ideally, the spec would use `readOnly` and `writeOnly` properties for these, but they were only added in OpenAPI 3, and the specs were migrated from OpenAPI 2, without adding any of the OpenAPI 3 features.

Not quite sure how to deal with such issues, given the implicit/ambiguous nature of such props. For now I guess it's best to just leave it optional.

**Comment by Karanjot786** [comment]
> This looks really good! :)
> 
> However, one major thing from looking at the nice summary that Sourcery provided: Are you sure you created the models for [TES v1.1](https://github.com/ga4gh/task-execution-schemas/blob/v1.1/openapi/task_execution_service.openapi.yaml) as previously discussed? There are a few minor inconsistencies with the models, but maybe it's because you have used an older version?
> 
> Here's a list of the issues I found:
> 
>     * `TESState` (from `tesState`):
>       
>       * `CANCELING` is missing (added in v1.1)
>       * `PREEMPTED` is missing (added in v1.1)
> 
>     * `TESInput` (from `tesInput`):
>       
>       * `type` is optional for creating tasks (requirement dropped in v1.1, I believe), but should be filled in once that information becomes available; probably not enforcable on the validation side
> 
>     * `TESOutput` (from `tesOutput`):
>       
>       * `path_prefix` is missing (added in v1.1); note that it is required if `TESOutput.path` contains wildcards; requires custom Pydantic validator
>       * `type` as above
> 
>     * `TESExecutor` (from `tesExecutor`):
>       
>       * `ignore_error` is missing (added in v1.1)
> 
>     * `TESData` (from `tesTask`):
>       
>       * Only `executors` is required, `inputs` and `outputs` are not
> 
> 
> Apart from that, a few minor comments:
> 
>     * Check the docstrings and make sure they are consistent with Google's docstring style and your other docstrings in the codebase (see individual comment for more info)
> 
>     * Address the open Sourcery comments

Thank you for your positive feedback and for taking the time to review my PR. I appreciate your thorough and detailed comments.

**TES Version Alignment**

You are correct; I inadvertently based the models on an older version of the TES schema. I apologize for this oversight. I will update the models to align with TES v1.1 as we talked about before. Specifically, I will address the following:

`TESState:`
        Add the missing `CANCELING` and `PREEMPTED` states introduced in v1.1.

`TESInput:`
        Update the `type` field to be optional during task creation, acknowledging that it may be filled in once the information becomes available.

`TESOutput:`
        Add the `path_prefix` field, which is required if the `path` contains wildcards, and implement a custom Pydantic validator to handle this scenario.
        Make the `type` field optional, consistent with v1.1.

`TESExecutor:`
        Include the `ignore_error` field added in v1.1.

`TESData:`
        Adjust the required fields so only `executors` are required, making `inputs` and `outputs` optional.

File: crategen/models/tes_models.py (hunk 1 above)
Comment (Karanjot786) [reply crategen/models/tes_models.py:37]: I will revise the docstrings throughout `tes_models.py` to ensure they conform to the Google-style docstring guidelines used in our codebase. This includes:

- **Consistency:** Making sure the docstrings are consistent with other parts of the codebase.

- **Style:** Removing any emphasis (e.g., bold or italics) from section headers and attribute names.

- **Type Annotations:** Ensuring that types are specified in the code via type hints and not repeated in the docstrings.

File: crategen/models/tes_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/models/tes_models.py:233]: I noticed the mention of `status` in the docstring of the `TESTaskLog` class, which is not present in the class attributes. Based on your feedback and the TES v1.1 schema, I will remove `status` from the docstring to avoid confusion.

Regarding the `state` field, I will ensure it is correctly implemented in `TESData` and marked as optional since it is set by the service and not by the user.

File: tests/unit/test_tes_models.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        volumes=["/data"],
+        tags={"project": "CrateGen"},
+    )
+    assert tes_data.id == "task-123"
+    assert tes_data.inputs[0].url == "https://raw.githubusercontent.com/elixir-cloud-aai/CrateGen/refs/heads/main/README.md"
+    assert tes_data.outputs[0].path == "/data/output/LICENSE"
+    assert tes_data.executors[0].image == "python:3.8-slim"
+    assert tes_data.resources.cpu_cores == EXPECTED_CPU_CORES
+    assert tes_data.volumes == ["/data"]
+    assert tes_data.tags == {"project": "CrateGen"}
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply tests/unit/test_tes_models.py:195]: I agree that the `test_tes_data_valid` function is quite extensive. I will refactor it by creating separate, focused tests for each subcomponent (`inputs`, `outputs`, `executors`, `resources`). This will improve test readability and maintainability.

File: tests/unit/test_tes_models.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+def test_tes_data_missing_required_fields():
+    """Test TESData model missing required fields (should raise ValidationError)."""
+    with pytest.raises(ValidationError) as exc_info:
+        TESData(
+            inputs=[],
+            outputs=[],
+            executors=[],
+        )
+    assert "field required" in str(exc_info.value)
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply tests/unit/test_tes_models.py:206]: I will enhance the `test_tes_data_missing_required_fields` test by asserting which specific fields are reported as missing. This will make the test more robust and informative by checking for precise validation errors.

**Comment by Karanjot786** [comment]
## Objective

- Introduce Pydantic models for Task Execution Services (TES) conforming to the GA4GH TES v1.1 schema.
- Provide comprehensive unit tests to ensure model integrity and functionality.
- Update docstrings to conform to Google style guidelines.
- Resolve all test errors and ensure code quality.

## Changes

### Updated Models:
- **`crategen/models/tes_models.py`**:
    - Added missing states `CANCELING` and `PREEMPTED` to `TESState`.
    - Made `type` field optional in `TESInput` and `TESOutput`.
    - Added `path_prefix` to `TESOutput` with appropriate validation.
    - Included `ignore_error` field in `TESExecutor`.
    - Adjusted `TESData` to make `inputs` and `outputs` optional, with only `executors` being required.

### Updated Unit Tests:
- **`tests/unit/test_tes_models.py`**:
    - Added tests for new fields and validators.
    - Refactored tests for better readability and coverage.
    - Ensured all tests pass successfully.

### Documentation:
- Revised docstrings in `tes_models.py` to align with Google style guidelines.
- Removed emphasis and type information from docstrings.
- Ensured consistency across all models.

### Linting and Testing:
- Resolved all linting issues using Ruff.
- Confirmed that all unit tests pass.

File: tests/unit/test_tes_models.py
[the remark this answers is not in the document]
Comment (Karanjot786) [reply tests/unit/test_tes_models.py:1]: Hi @SalihuDickson,

Thank you for your feedback and for highlighting the importance of focusing our tests on our own logic rather than external packages. I now understand that when writing tests for the models, I should concentrate on testing the custom validators and any additional logic we've implemented, rather than the inherent behavior of Pydantic models.

I appreciate your offer to rewrite the tests to provide an example. That will be incredibly helpful for me to learn the proper approach to structuring these tests. I'll make sure to study your example thoroughly and apply the same principles in future test cases.

Thank you again for your guidance and support.

**Comment by SalihuDickson** [comment]
@uniqueg, I think this is just about ready to merge, maybe you want to do a quick read through of the recent changes.

**Review by salihudickson — APPROVED** [review]

**Comment by SalihuDickson** [comment]
hi @uniqueg, please take a look at this whenever you have the time.

**Comment by Karanjot786** [comment]
Hi @SalihuDickson, I've taken a look at the updates you've made, and they look great. I appreciate the example you've provided, and I will implement the same method in the WES models as well.

**Review by salihudickson — APPROVED** [review]

**Comment by SalihuDickson** [comment]
hey @uniqueg, just a quick follow up on my earlier reminder. Need your approval to merge :)

**Review by uniqueg — APPROVED** [review]
Nice - good job @SalihuDickson. Good to merge, but please remember that you indicate @Karanjot786 as co-author here.

**EDIT:** Actually, it'll be the other way around - you'll be co-author - and this is automatically picked up, so you are good to merge, @SalihuDickson.

**Comment by Karanjot786** [comment]
Thanks, @uniqueg! Understood. I’ll proceed with the merge now.
